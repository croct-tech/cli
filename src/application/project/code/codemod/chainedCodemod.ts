import {Codemod, CodemodOptions, ResultCode} from '@/application/project/code/codemod/codemod';

export class ChainedCodemod<I, O extends CodemodOptions> implements Codemod<I, O> {
    private readonly codemods: Array<Codemod<I, O>>;

    public constructor(...codemods: Array<Codemod<I, O>>) {
        this.codemods = codemods;
    }

    public async apply(input: I, options?: O): Promise<ResultCode<I>> {
        let result: I = input;
        let modified = false;

        for (const codemod of this.codemods) {
            const currentResult = await codemod.apply(result, options);

            result = currentResult.result;
            modified = modified || currentResult.modified;
        }

        return {
            modified: modified,
            result: result,
        };
    }
}
