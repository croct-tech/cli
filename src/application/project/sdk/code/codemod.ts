export type ResultCode<T> = {
    modified: boolean,
    result: T,
};

export type CodemodOptions = Record<string, any>;

export interface Codemod<I, O extends CodemodOptions = CodemodOptions> {
    apply(input: I, options?: O): Promise<ResultCode<I>>;
}

export class CodemodError extends Error {
    public constructor(message: string) {
        super(message);

        Object.setPrototypeOf(this, CodemodError.prototype);
    }
}

export class MalformedCodeError extends CodemodError {
    public constructor(message: string) {
        super(message);

        Object.setPrototypeOf(this, MalformedCodeError.prototype);
    }
}
