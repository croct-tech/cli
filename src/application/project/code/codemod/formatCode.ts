import {Codemod, CodemodOptions, ResultCode} from '@/application/project/code/codemod/codemod';
import {CodeFormatter} from '@/application/project/code/formatter/formatter';

export class FormatCode<O extends CodemodOptions> implements Codemod<string, O> {
    private readonly codemod: Codemod<string, O>;

    private readonly formatter: CodeFormatter;

    public constructor(codemod: Codemod<string, O>, formatter: CodeFormatter) {
        this.codemod = codemod;
        this.formatter = formatter;
    }

    public async apply(input: string, options?: O): Promise<ResultCode<string>> {
        const result = await this.codemod.apply(input, options);

        if (result.modified) {
            await this.formatter.format([input]);
        }

        return result;
    }
}
