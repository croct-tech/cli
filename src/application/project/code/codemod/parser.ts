import {parse as babelParse} from '@babel/parser';
import {File} from '@babel/types';
import {MalformedCodeError} from '@/application/project/code/codemod/codemod';

export type Language = 'jsx' | 'typescript';

export function parse(source: string, languages: Language[]): File {
    try {
        return babelParse(source, {
            sourceType: 'module',
            plugins: languages,
        });
    } catch {
        throw new MalformedCodeError('The source code contains syntax errors.');
    }
}
