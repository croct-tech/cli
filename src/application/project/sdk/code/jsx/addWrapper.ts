/* eslint-disable no-param-reassign -- False positives */
import * as t from '@babel/types';
import traverse from '@babel/traverse';
import {traverseFast} from '@babel/types';
import {ResultCode, Codemod, CodemodOptions} from '@/application/project/sdk/code/codemod';
import {addImport} from '@/application/project/sdk/code/javascript/addImport';

type ComponentDeclaration = t.VariableDeclarator | t.FunctionDeclaration;
type DeclarationKind = t.ExportDefaultDeclaration['declaration'];
type ExpressionKind = t.Expression;
type JsxKind = t.JSXElement['children'][number];

type TargetChildren = {
    parent: t.JSXElement,
    index: number,
};

type PropertyTypes = {
    env: {
        name: string,
    },
    literal: {
        value: string | number | boolean | null,
    },
};

type PropertyType = {
    [K in keyof PropertyTypes]: PropertyTypes[K] & {type: K}
}[keyof PropertyTypes];

export type WrapperOptions<O extends CodemodOptions = CodemodOptions> = {
    wrapper: {
        component: string,
        module: string,
        props?: Record<string, PropertyType>,
    },
    targets?: {
        variable?: string,
        component?: string,
    },
    fallbackToNamedExports?: boolean,
    fallbackCodemod?: Codemod<t.File, O>,
};

enum Transformation {
    APPLIED,
    NOT_APPLIED,
    ALREADY_APPLIED,
}

type WrapperInsertion = {
    result: Transformation,
    node: t.Expression,
};

/**
 * Wraps the exported component with a wrapper component.
 *
 * This transformation is useful for adding wrapper elements such as context providers.
 *
 * It attempts to wrap the default export first, and if not found, it can optionally
 * wrap named exports that return JSX elements depending on the configuration.
 */
export class AddWrapper<O extends CodemodOptions = CodemodOptions> implements Codemod<t.File> {
    private readonly options: WrapperOptions<O>;

    public constructor(options: WrapperOptions<O>) {
        this.options = options;
    }

    public apply(input: t.File, options: O): Promise<ResultCode<t.File>> {
        const ast = t.cloneNode(input, true);

        const componentImport = addImport(ast, {
            type: 'value',
            moduleName: this.options.wrapper.module,
            importName: this.options.wrapper.component,
        });

        const component = componentImport.localName ?? this.options.wrapper.component;
        const namedExports: t.ExportNamedDeclaration[] = [];

        let result = Transformation.NOT_APPLIED;

        traverse(ast, {
            ExportDefaultDeclaration: path => {
                // Wrap the default export
                result = this.wrapDeclaration(path.node.declaration, component, ast);

                return path.stop();
            },
            ExportNamedDeclaration: path => {
                // Collect named exports for fallback wrapping
                namedExports.push(path.node);

                return path.skip();
            },
        });

        if (result === Transformation.NOT_APPLIED && this.options?.fallbackToNamedExports === true) {
            // If the default export was not found, attempt to wrap named exports
            for (const namedExport of namedExports) {
                if (t.isFunctionDeclaration(namedExport.declaration)) {
                    // export function Component() { ... }
                    result = this.wrapBlockStatement(namedExport.declaration.body, component, ast);

                    if (result === Transformation.APPLIED) {
                        break;
                    }
                }

                if (t.isVariableDeclaration(namedExport.declaration)) {
                    // export const Component = function() { ... };
                    const {declarations} = namedExport.declaration;

                    // Find the first variable declaration that is a variable,
                    // arrow function, or function expression
                    const modified = declarations.some(declaration => {
                        if (
                            t.isIdentifier(declaration.init)
                            || t.isArrowFunctionExpression(declaration.init)
                            || t.isFunctionExpression(declaration.init)
                        ) {
                            return this.wrapDeclaration(declaration.init, component, ast) === Transformation.APPLIED;
                        }

                        return false;
                    });

                    if (modified) {
                        result = Transformation.APPLIED;

                        break;
                    }
                }

                // export {Component as SomeComponent};
                for (const specifier of namedExport.specifiers ?? []) {
                    if (t.isExportSpecifier(specifier)) {
                        const declaration = this.findComponentDeclaration(ast, specifier.local.name);

                        if (
                            declaration !== null
                            && t.isVariableDeclarator(declaration)
                            && t.isExpression(declaration.init)
                        ) {
                            result = this.wrapDeclaration(declaration.init, component, ast);

                            if (result === Transformation.APPLIED) {
                                break;
                            }
                        }
                    }
                }

                if (result === Transformation.APPLIED) {
                    break;
                }
            }
        }

        const fallbackCodemod = this.options?.fallbackCodemod;

        if (result === Transformation.NOT_APPLIED && fallbackCodemod !== undefined) {
            return fallbackCodemod.apply(input, options);
        }

        return Promise.resolve({
            modified: result === Transformation.APPLIED,
            result: result === Transformation.APPLIED ? ast : input,
        });
    }

    /**
     * Wraps the declaration with the wrapper component.
     *
     * The declaration can be a function declaration, arrow function,
     * or variable declaration.
     *
     * @param node The declaration to wrap.
     * @param component The name of the component to use as the wrapper.
     * @param ast The AST representing the source code.
     * @return the result of the transformation.
     */
    private wrapDeclaration(node: DeclarationKind, component: string, ast: t.File): Transformation {
        if (t.isArrowFunctionExpression(node)) {
            if (t.isBlockStatement(node.body)) {
                return this.wrapBlockStatement(node.body, component, ast);
            }

            const insertion = this.insertWrapper(node.body, component, ast);

            if (insertion.result !== Transformation.APPLIED) {
                return insertion.result;
            }

            node.body = insertion.node;

            return Transformation.APPLIED;
        }

        if (t.isFunctionExpression(node) || t.isFunctionDeclaration(node)) {
            return this.wrapBlockStatement(node.body, component, ast);
        }

        if (t.isIdentifier(node)) {
            // Find the declaration of the component by name
            const declaration = this.findComponentDeclaration(ast, node.name);

            if (declaration !== null) {
                if (t.isVariableDeclarator(declaration)) {
                    const initializer = declaration.init ?? null;

                    if (initializer !== null) {
                        return this.wrapDeclaration(initializer, component, ast);
                    }
                } else {
                    return this.wrapBlockStatement(declaration.body, component, ast);
                }
            }
        }

        return Transformation.NOT_APPLIED;
    }

    /**
     * Wraps the block statement with the wrapper component.
     *
     * @param node The block statement to wrap
     * @param component The name of the component to use as the wrapper.
     * @param ast The AST representing the source code.
     * @return the result of the transformation.
     */
    private wrapBlockStatement(node: t.BlockStatement, component: string, ast: t.File): Transformation {
        const returnStatement = AddWrapper.findReturnStatement(node);
        const argument = returnStatement?.argument ?? null;

        if (returnStatement !== null && argument !== null) {
            const insertion = this.insertWrapper(argument, component, ast);

            if (insertion.result !== Transformation.APPLIED) {
                return insertion.result;
            }

            returnStatement.argument = t.parenthesizedExpression(insertion.node);

            return Transformation.APPLIED;
        }

        return Transformation.NOT_APPLIED;
    }

    /**
     * Wraps the target element with the wrapper component.
     *
     * The given node may not be the target element itself, but a parent
     * that contains the target element. For example, the target element
     * may be a `{children}` expression inside a JSX element.
     *
     * @param node The node containing the target element.
     * @param component The name of the component to use as the wrapper.
     * @param ast The AST representing the source code.
     * @return The result of the transformation.
     */
    private insertWrapper(node: ExpressionKind, component: string, ast: t.File): WrapperInsertion {
        if (this.containsElement(node, component)) {
            return {
                result: Transformation.ALREADY_APPLIED,
                node: node,
            };
        }

        const target = this.findTargetChildren(ast, node);

        if (target !== null) {
            const {parent, index} = target;

            const children = [...parent.children ?? []];
            const child = children.splice(index, 1)[0] as JsxKind;

            target.parent.children = children.length === 0
                // If there is only one child, replace it with the wrapper
                ? [
                    t.jsxText('\n'),
                    this.wrapElement(child, component),
                    t.jsxText('\n'),
                ]
                // Otherwise, insert the wrapper at the target index
                : [
                    ...children.slice(0, index),
                    this.wrapElement(child, component),
                    ...children.slice(index),
                ];

            return {
                result: Transformation.APPLIED,
                node: node,
            };
        }

        if (
            t.isJSXText(node)
            || t.isJSXExpressionContainer(node)
            || t.isJSXSpreadChild(node)
            || t.isJSXElement(node)
            || t.isJSXFragment(node)
        ) {
            // Wrap the whole element if the target is not found
            return {
                result: Transformation.APPLIED,
                node: this.wrapElement(node, component),
            };
        }

        return {
            result: Transformation.NOT_APPLIED,
            node: node,
        };
    }

    /**
     * Wraps the given JSX element with the wrapper component.
     *
     * @param node The JSX element to wrap.
     * @param name The name of the component to use as the wrapper.
     * @return The wrapped JSX element.
     */
    private wrapElement(node: JsxKind, name?: string): t.JSXElement {
        return t.jsxElement(
            t.jsxOpeningElement(
                t.jsxIdentifier(name ?? this.options.wrapper.component),
                this.getProviderProps(),
            ),
            t.jsxClosingElement(
                t.jsxIdentifier(name ?? this.options.wrapper.component),
            ),
            [
                t.jsxText('\n'),
                node,
                t.jsxText('\n'),
            ],
        );
    }

    /**
     * Returns the props for the wrapper component.
     *
     * @return The list of JSX attributes for the wrapper component.
     */
    private getProviderProps(): t.JSXAttribute[] {
        const attributes: t.JSXAttribute[] = [];
        const props = this.options.wrapper.props ?? {};

        for (const [key, value] of Object.entries(props)) {
            attributes.push(
                t.jsxAttribute(
                    t.jsxIdentifier(key),
                    t.jsxExpressionContainer(
                        value.type === 'env'
                            ? t.memberExpression(
                                t.memberExpression(
                                    t.identifier('process'),
                                    t.identifier('env'),
                                ),
                                t.identifier(value.name),
                            )
                            : AddWrapper.getLiteralValue(value.value),
                    ),
                ),
            );
        }

        return attributes;
    }

    /**
     * Returns the literal value as an AST node.
     *
     * @param value The value to represent as an AST node.
     * @return The AST node representing the value.
     */
    private static getLiteralValue(
        value: string | number | boolean | null,
    ): t.StringLiteral | t.NumericLiteral | t.BooleanLiteral | t.NullLiteral {
        if (typeof value === 'string') {
            return t.stringLiteral(value);
        }

        if (typeof value === 'number') {
            return t.numericLiteral(value);
        }

        if (typeof value === 'boolean') {
            return t.booleanLiteral(value);
        }

        return t.nullLiteral();
    }

    /**
     * Determines if the element contains the specified component.
     *
     * @param parent The parent element to search for the component.
     * @param name The name of the component to find.
     * @return true if the element contains the component, false otherwise.
     */
    private containsElement(parent: t.Node, name: string): boolean {
        let contains = false;

        traverseFast(parent, node => {
            if (
                !contains
                && t.isJSXOpeningElement(node)
                && t.isJSXIdentifier(node.name)
                && node.name.name === name
            ) {
                contains = true;
            }
        });

        return contains;
    }

    /**
     * Finds the declaration of the component by name.
     *
     * @param ast The AST representing the source code.
     * @param name The name of the variable or function to find.
     * @return The declaration of the component or null if not found.
     */
    private findComponentDeclaration(ast: t.File, name: string): ComponentDeclaration | null {
        let componentDeclaration: ComponentDeclaration | null = null;

        traverse(ast, {
            VariableDeclaration: path => {
                if (!t.isProgram(path.parent)) {
                    return path.skip();
                }

                for (const declaration of path.node.declarations) {
                    if (
                        t.isVariableDeclarator(declaration)
                        && t.isIdentifier(declaration.id)
                        && declaration.id.name === name
                    ) {
                        componentDeclaration = declaration;

                        return path.stop();
                    }
                }

                return path.skip();
            },
            FunctionDeclaration: path => {
                if (!t.isProgram(path.parent)) {
                    return path.skip();
                }

                const {id} = path.node;

                if (t.isIdentifier(id) && id.name === name) {
                    componentDeclaration = path.node;

                    return path.stop();
                }

                return path.skip();
            },
        });

        return componentDeclaration;
    }

    /**
     * Finds the target children of wrapper component.
     *
     * @param ast The AST representing the source code.
     * @param element The element to search for the target.
     * @return The parent element and index of the target children or null if not found.
     */
    private findTargetChildren(ast: t.File, element: t.Node): TargetChildren | null {
        let insertionPoint: TargetChildren | null = null;
        const {options} = this;

        if (options.targets?.variable === undefined && options.targets?.component === undefined) {
            // If no targets are specified, wrap the whole element
            return null;
        }

        traverse(ast, {
            enter: function enter(path) {
                const {node} = path;

                if (node !== element) {
                    return;
                }

                path.traverse({
                    JSXExpressionContainer: nestedPath => {
                        const {expression} = nestedPath.node;

                        if (
                            options.targets?.variable !== undefined
                            && t.isIdentifier(expression)
                            && expression.name === options.targets.variable
                        ) {
                            const parent = nestedPath.parent as t.JSXElement;

                            insertionPoint = {
                                parent: parent,
                                index: parent.children.indexOf(nestedPath.node),
                            };

                            return nestedPath.stop();
                        }
                    },
                    JSXElement: nestedPath => {
                        const {openingElement} = nestedPath.node;

                        if (
                            options.targets?.component !== undefined
                            && t.isJSXOpeningElement(openingElement)
                            && t.isJSXIdentifier(openingElement.name)
                            && openingElement.name.name === options.targets.component
                        ) {
                            if (nestedPath.parent !== null) {
                                const parent = nestedPath.parent as t.JSXElement;

                                insertionPoint = {
                                    parent: parent,
                                    index: parent.children.indexOf(nestedPath.node),
                                };
                            }
                        }
                    },
                });

                return path.stop();
            },
        });

        return insertionPoint;
    }

    /**
     * Finds the return statement in the block statement.
     *
     * @param body The block statement to search for the return statement.
     * @return The return statement or null if not found.
     */
    private static findReturnStatement(body: t.BlockStatement): t.ReturnStatement | null {
        let returnStatement: t.ReturnStatement | null = null;

        traverseFast(body, node => {
            if (returnStatement === null && t.isReturnStatement(node)) {
                returnStatement = node;
            }
        });

        return returnStatement;
    }
}
