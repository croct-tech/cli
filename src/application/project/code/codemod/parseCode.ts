import {Node} from '@babel/types';
import generate from '@babel/generator';
import {Codemod, CodemodOptions, ResultCode} from '@/application/project/code/codemod/codemod';
import {Language, parse} from '@/application/project/code/codemod/parser';

export type Configuration<O extends CodemodOptions> = {
    codemod: Codemod<Node, O>,
    languages: Language[],
};

export class ParseCode<O extends CodemodOptions> implements Codemod<string, O> {
    private readonly codemod: Codemod<Node, O>;

    private readonly languages: Language[];

    public constructor(configuration: Configuration<O>) {
        this.codemod = configuration.codemod;
        this.languages = configuration.languages;
    }

    public async apply(input: string, options?: O): Promise<ResultCode<string>> {
        const result = await this.codemod.apply(parse(input, this.languages), options);

        if (!result.modified) {
            return {
                modified: false,
                result: input,
            };
        }

        return {
            modified: true,
            result: generate(
                result.result,
                {
                    compact: false,
                    retainFunctionParens: true,
                },
            ).code,
        };
    }
}
