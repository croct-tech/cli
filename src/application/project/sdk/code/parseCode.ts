import type {namedTypes as Ast} from 'ast-types';
import {print} from 'recast';
import {Codemod, CodemodOptions, ResultCode} from '@/application/project/sdk/code/codemod';
import {Language, parse} from '@/application/project/sdk/code/parser';

export type Configuration<O extends CodemodOptions> = {
    codemod: Codemod<Ast.Node, O>,
    languages: Language[],
};

export class ParseCode<O extends CodemodOptions> implements Codemod<string, O> {
    private readonly codemod: Codemod<Ast.Node, O>;

    private readonly languages: Language[];

    public constructor(configuration: Configuration<O>) {
        this.codemod = configuration.codemod;
        this.languages = configuration.languages;
    }

    public async apply(input: string, options?: O): Promise<ResultCode<string>> {
        const result = await this.codemod.apply(parse(input, this.languages), options);

        if (!result.modified) {
            return {
                modified: false,
                result: input,
            };
        }

        return {
            modified: true,
            result: print(result.result).code,
        };
    }
}
