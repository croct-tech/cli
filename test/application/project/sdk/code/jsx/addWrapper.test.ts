import {resolve} from 'path';
import {addComment, File} from '@babel/types';
import {AddWrapper, WrapperConfiguration} from '@/application/project/code/codemod/jsx/addWrapper';
import {loadFixtures} from '../fixtures';
import {ParseCode} from '@/application/project/code/codemod/parseCode';
import {Codemod} from '@/application/project/code/codemod/codemod';

describe('AddWrapper', () => {
    const defaultOptions: WrapperConfiguration = {
        wrapper: {
            module: '@croct/plug-react',
            component: 'CroctProvider',
        },
    };

    const scenarios = loadFixtures<WrapperConfiguration>(
        resolve(__dirname, '../fixtures/jsx-wrapper'),
        defaultOptions,
        // Custom options
        {
            'targetComponent.tsx': {
                targets: {
                    component: 'Component',
                },
            },
            'targetChildren.tsx': {
                targets: {
                    variable: 'children',
                },
            },
            'targetChildrenAliasedImport.tsx': {
                targets: {
                    variable: 'children',
                },
            },
            'namedExportArrowFunctionWithBody.tsx': {
                fallbackToNamedExports: true,
            },
            'namedExportArrowFunction.tsx': {
                fallbackToNamedExports: true,
            },
            'namedExportFunctionDeclaration.tsx': {
                fallbackToNamedExports: true,
            },
            'namedExportFunctionExpression.tsx': {
                fallbackToNamedExports: true,
            },
            'namedSpecifiedExport.tsx': {
                fallbackToNamedExports: true,
            },
            'namedExportUnrelated.tsx': {
                fallbackToNamedExports: true,
            },
            'defaultExportFunctionReference.tsx': {
                targets: {
                    variable: 'children',
                },
            },
            'defaultExportArrowFunctionReference.tsx': {
                targets: {
                    variable: 'children',
                },
            },
            'defaultExportFunctionExpressionReference.tsx': {
                targets: {
                    variable: 'children',
                },
            },
            'providerObjectMemberProp.tsx': {
                wrapper: {
                    ...defaultOptions.wrapper,
                    props: {
                        appId: {
                            type: 'reference',
                            path: ['process', 'env', 'REACT_APP_CROCT_APP_ID'],
                        },
                    },
                },
            },
            'providerVariableProp.tsx': {
                wrapper: {
                    ...defaultOptions.wrapper,
                    props: {
                        appId: {
                            type: 'reference',
                            path: ['CROCT_APP_ID'],
                        },
                    },
                },
            },
            'providerTernaryProp.tsx': {
                wrapper: {
                    ...defaultOptions.wrapper,
                    props: {
                        booleanProp: {
                            type: 'ternary',
                            condition: {
                                operator: '===',
                                left: {
                                    type: 'reference',
                                    path: ['process', 'env', 'NODE_ENV'],
                                },
                                right: {
                                    type: 'literal',
                                    value: 'production',
                                },
                            },
                            consequent: {
                                type: 'reference',
                                path: ['process', 'env', 'PRODUCTION_CROCT_APP_ID'],
                            },
                            alternate: {
                                type: 'reference',
                                path: ['process', 'env', 'STAGING_CROCT_APP_ID'],
                            },
                        },
                    },
                },
            },
            'providerLiteralProp.tsx': {
                wrapper: {
                    ...defaultOptions.wrapper,
                    props: {
                        booleanProp: {
                            type: 'literal',
                            value: true,
                        },
                        numberProp: {
                            type: 'literal',
                            value: 42,
                        },
                        stringProp: {
                            type: 'literal',
                            value: 'value',
                        },
                        nullProp: {
                            type: 'literal',
                            value: null,
                        },
                    },
                },
            },
        },
    );

    it.each(scenarios)('should correctly transform $name', async ({name, fixture, options}) => {
        const transformer = new ParseCode({
            languages: ['typescript', 'jsx'],
            codemod: new AddWrapper(options),
        });

        const output = await transformer.apply(fixture);

        expect(output.result).toMatchSnapshot(name);
    });

    it('should call the fallback codemod when the component is not found', async () => {
        const options = {test: true};

        const codemod: Codemod<File, typeof options> = {
            apply: jest.fn((input: File) => {
                const {body} = input.program;

                addComment(body[0], 'leading', ' This is a comment');

                return Promise.resolve({
                    modified: true,
                    result: input,
                });
            }),
        };

        const transformer = new ParseCode({
            languages: ['typescript', 'jsx'],
            codemod: new AddWrapper({
                ...defaultOptions,
                fallbackCodemod: codemod,
            }),
        });

        const {result} = await transformer.apply('export {Component} from \'./component\';', options);

        expect(codemod.apply).toHaveBeenCalledWith(expect.any(Object), options);

        expect(result).toEqual(
            '/* This is a comment*/\n'
            + 'export { Component } from \'./component\';',
        );
    });

    it('should not call the fallback codemod when the wrapper is already present', async () => {
        const codemod: Codemod<File> = {
            apply: jest.fn(),
        };

        const transformer = new ParseCode({
            languages: ['typescript', 'jsx'],
            codemod: new AddWrapper<{test: boolean}>({
                ...defaultOptions,
                fallbackCodemod: codemod,
            }),
        });

        const input = [
            "import {CroctProvider} from '@croct/plug-react';",
            'export default function Component() {',
            '  return <CroctProvider/>',
            '}',
        ].join('\n');

        const {result} = await transformer.apply(input);

        expect(codemod.apply).not.toHaveBeenCalled();

        expect(result).toEqual(input);
    });
});
