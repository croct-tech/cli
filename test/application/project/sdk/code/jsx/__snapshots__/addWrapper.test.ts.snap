// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddWrapper should correctly transform aliasedImport.tsx: aliasedImport.tsx 1`] = `
"import type { AppProps } from "next/app";
import { ReactElement } from "react";
import { CroctProvider as Provider } from "@croct/plug-react";
export default function App({
  Component,
  pageProps
}: AppProps): ReactElement {
  return (<Provider>
    <Component {...pageProps} />
  </Provider>);
}"
`;

exports[`AddWrapper should correctly transform defaultExport.tsx: defaultExport.tsx 1`] = `
"import { CroctProvider } from "@croct/plug-react";
import type { AppProps } from "next/app";
import { ReactElement } from "react";
export default function App({
  Component,
  pageProps
}: AppProps): ReactElement {
  return (<CroctProvider>
    <Component {...pageProps} />
  </CroctProvider>);
}"
`;

exports[`AddWrapper should correctly transform defaultExportAnonymousFunctionWithUnrelatedReturn.tsx: defaultExportAnonymousFunctionWithUnrelatedReturn.tsx 1`] = `
"export default function () {
    return null;
}
"
`;

exports[`AddWrapper should correctly transform defaultExportArrowFunctionReference.tsx: defaultExportArrowFunctionReference.tsx 1`] = `
"import { CroctProvider } from "@croct/plug-react";
const Layout = ({
  children
}) => {
  return (<>
        <main>
            <CroctProvider>
        {children}
      </CroctProvider>
        </main>
    </>);
};
export default Layout;"
`;

exports[`AddWrapper should correctly transform defaultExportArrowFunctionWithUnrelatedReturn.tsx: defaultExportArrowFunctionWithUnrelatedReturn.tsx 1`] = `
"export default () => null;
"
`;

exports[`AddWrapper should correctly transform defaultExportFunctionExpressionReference.tsx: defaultExportFunctionExpressionReference.tsx 1`] = `
"import { CroctProvider } from "@croct/plug-react";
const Layout = function ({
  children
}) {
  return (<>
        <main>
            <CroctProvider>
        {children}
      </CroctProvider>
        </main>
    </>);
};
export default Layout;"
`;

exports[`AddWrapper should correctly transform defaultExportFunctionReference.tsx: defaultExportFunctionReference.tsx 1`] = `
"import { CroctProvider } from "@croct/plug-react";
function Layout({
  children
}) {
  return (<>
        <main>
            <CroctProvider>
        {children}
      </CroctProvider>
        </main>
    </>);
}
export default Layout;"
`;

exports[`AddWrapper should correctly transform defaultExportFunctionWithoutReturn.tsx: defaultExportFunctionWithoutReturn.tsx 1`] = `
"export default function () {
}
"
`;

exports[`AddWrapper should correctly transform defaultExportIndirectReference.tsx: defaultExportIndirectReference.tsx 1`] = `
"import { CroctProvider } from "@croct/plug-react";
// Block scoped variable, unrelated to the default export
{
  const Layout = () => {};
  const Reference = Layout;
}
const Layout = ({
  children
}) => {
  return (<CroctProvider>
    <>
        <main>
            {children}
        </main>
    </>
  </CroctProvider>);
};
const Reference = Layout;
export default Reference;"
`;

exports[`AddWrapper should correctly transform defaultExportUnrelated.tsx: defaultExportUnrelated.tsx 1`] = `
"export default 1;
"
`;

exports[`AddWrapper should correctly transform defaultExportWrapped.tsx: defaultExportWrapped.tsx 1`] = `
"import type {AppProps} from "next/app";
import {ReactElement} from "react";
import {CroctProvider} from "@croct/plug-react";

export default function App({Component, pageProps}: AppProps): ReactElement {
    return (
        <CroctProvider>
            <Component {...pageProps} />;
        </CroctProvider>
    );
}
"
`;

exports[`AddWrapper should correctly transform defaultExportWrappedAlias.tsx: defaultExportWrappedAlias.tsx 1`] = `
"import type {AppProps} from "next/app";
import {ReactElement} from "react";
import {CroctProvider as Provider} from "@croct/plug-react";

export default function App({Component, pageProps}: AppProps): ReactElement {
    return (
        <Provider>
            <Component {...pageProps} />;
        </Provider>
    );
}
"
`;

exports[`AddWrapper should correctly transform namedExportArrowFunction.tsx: namedExportArrowFunction.tsx 1`] = `
"import { CroctProvider } from "@croct/plug-react";
const UnrelatedComponent = ({
  Component,
  pageProps
}) => {
  return <Component {...pageProps} />;
};
export const App = ({
  Component,
  pageProps
}) => <CroctProvider>
  <Component {...pageProps} />
</CroctProvider>;"
`;

exports[`AddWrapper should correctly transform namedExportArrowFunctionWithBody.tsx: namedExportArrowFunctionWithBody.tsx 1`] = `
"import { CroctProvider } from "@croct/plug-react";
const UnrelatedComponent = ({
  Component,
  pageProps
}) => {
  return <Component {...pageProps} />;
};
export const App = ({
  Component,
  pageProps
}) => {
  return (<CroctProvider>
    <Component {...pageProps} />
  </CroctProvider>);
};"
`;

exports[`AddWrapper should correctly transform namedExportArrowFunctionWrapped.tsx: namedExportArrowFunctionWrapped.tsx 1`] = `
"import {CroctProvider} from "@croct/plug-react";

const UnrelatedComponent = ({Component, pageProps}) => {
    return <Component {...pageProps} />;
}

export const App = ({Component, pageProps}) => (
    <CroctProvider>
        <Component {...pageProps} />
    </CroctProvider>
);
"
`;

exports[`AddWrapper should correctly transform namedExportFunctionDeclaration.tsx: namedExportFunctionDeclaration.tsx 1`] = `
"import { CroctProvider } from "@croct/plug-react";
const UnrelatedComponent = ({
  Component,
  pageProps
}) => {
  return <Component {...pageProps} />;
};
export function App({
  Component,
  pageProps
}) {
  return (<CroctProvider>
    <Component {...pageProps} />
  </CroctProvider>);
}"
`;

exports[`AddWrapper should correctly transform namedExportFunctionExpression.tsx: namedExportFunctionExpression.tsx 1`] = `
"import { CroctProvider } from "@croct/plug-react";
const UnrelatedComponent = ({
  Component,
  pageProps
}) => {
  return <Component {...pageProps} />;
};
export const App = function ({
  Component,
  pageProps
}) {
  return (<CroctProvider>
    <Component {...pageProps} />
  </CroctProvider>);
};"
`;

exports[`AddWrapper should correctly transform namedExportUnrelated.tsx: namedExportUnrelated.tsx 1`] = `
"export const foo = 1;
"
`;

exports[`AddWrapper should correctly transform namedSpecifiedExport.tsx: namedSpecifiedExport.tsx 1`] = `
"import { CroctProvider } from "@croct/plug-react";
// Block scoped variable, unrelated to the default export
{
  function Layout() {}
  const Reference = Layout;
}
function unrelated() {}
const Layout = ({
  children
}) => {
  return (<CroctProvider>
    <>
        <main>
            {children}
        </main>
    </>
  </CroctProvider>);
};
const Reference = Layout;
export { Reference as Layout };"
`;

exports[`AddWrapper should correctly transform parenthesizedReturn.tsx: parenthesizedReturn.tsx 1`] = `
"import { CroctProvider } from "@croct/plug-react";
export default function App({
  Component,
  pageProps
}) {
  return (<CroctProvider>
    <div>
            <section>
                <Component {...pageProps} />
            </section>
        </div>
  </CroctProvider>);
}"
`;

exports[`AddWrapper should correctly transform providerLiteralProp.tsx: providerLiteralProp.tsx 1`] = `
"import { CroctProvider } from "@croct/plug-react";
export default function Layout({
  children
}) {
  return (<CroctProvider booleanProp={true} numberProp={42} stringProp={"value"} nullProp={null}>
    <>
            <main>{children}</main>
        </>
  </CroctProvider>);
}"
`;

exports[`AddWrapper should correctly transform providerReferenceProp.tsx: providerReferenceProp.tsx 1`] = `
"import { CroctProvider } from "@croct/plug-react";
export default function Layout({
  children
}) {
  return (<CroctProvider appId={process.env.REACT_APP_CROCT_APP_ID}>
    <>
            <main>{children}</main>
        </>
  </CroctProvider>);
}"
`;

exports[`AddWrapper should correctly transform providerTernaryProp.tsx: providerTernaryProp.tsx 1`] = `
"import { CroctProvider } from "@croct/plug-react";
export default function Layout({
  children
}) {
  return (<CroctProvider booleanProp={process.env.NODE_ENV === "production" ? process.env.PRODUCTION_CROCT_APP_ID : process.env.STAGING_CROCT_APP_ID}>
    <>
            <main>{children}</main>
        </>
  </CroctProvider>);
}"
`;

exports[`AddWrapper should correctly transform providerVariableProp.tsx: providerVariableProp.tsx 1`] = `
"import { CroctProvider } from "@croct/plug-react";
export default function Layout({
  children
}) {
  return (<CroctProvider appId={CROCT_APP_ID}>
    <>
            <main>{children}</main>
        </>
  </CroctProvider>);
}"
`;

exports[`AddWrapper should correctly transform targetChildren.tsx: targetChildren.tsx 1`] = `
"import { CroctProvider } from "@croct/plug-react";
export default function Layout({
  children
}) {
  return (<html lang="en">
            <body>
                {/* This is a comment */}
                <CroctProvider>
        {children}
      </CroctProvider>
            </body>
        </html>);
}"
`;

exports[`AddWrapper should correctly transform targetChildrenAliasedImport.tsx: targetChildrenAliasedImport.tsx 1`] = `
"import { CroctProvider as Provider } from "@croct/plug-react";
export default function Layout({
  children
}) {
  return (<html lang="en">
            <body>
                {/* This is a comment */}
                <Provider>
        {children}
      </Provider>
            </body>
        </html>);
}"
`;

exports[`AddWrapper should correctly transform targetComponent.tsx: targetComponent.tsx 1`] = `
"import { CroctProvider } from "@croct/plug-react";
export default function App({
  Component,
  pageProps
}) {
  return (<div>
        <section>
            <CroctProvider>
        <Component {...pageProps} />
      </CroctProvider>
        </section>
    </div>);
}"
`;
